# 05-5 내장 함수
<hr style="height: 1px;">

이미 만들어진 것을 다시 만드는 것은 불필요한 행동이다(Don't reinvest the wheel.) <br>
파이썬 배포본에 함께 들어 있는 파이썬 라이브러리를 살펴보는 것은 매우 중요. <br>
라이브러리를 살펴보기 전에 파이썬 내장 함수를 먼저 보자. 파이썬 내장 함수는 외부 모듈과 달리 import가 필요하지 않음.<br>


## 1. abs
<hr>
abs(x)는 숫자의 절대값을 돌려주는 함수

In [1]:
abs(-3)

3

In [2]:
abs(1.2)

1.2

## 2. all
<hr>
all(x)는 반복 가능한(iterable) 자료형 x를 입력 인수로 받으며 이 x가 모두 참이면 True, 거짓이 하나라도 있으면 False를 반환<br>

> 반복 가능한 자료형이란 for문으로 그 값을 출력할 수 있는 것을 의미. 리스트, 튜플, 문자열, 딕셔너리, 집합 등

리스트 자료형 [1, 2, 3]은 모든 요소가 참이므로 True

In [3]:
all([1, 2, 3])

True

리스트 자료형 [1, 2, 3, 0] 중에서 요소 0은 거짓이므로 False

In [5]:
all([1, 2, 3, 0])

False

## 3. any
<hr>
any(x)는 x 중 하나라도 참이 있으면 True를 돌려주고, x가 모두 거짓일 때에만 False를 돌려줌. all(x)의 반대.

리스트 자료형 [1, 2, 3, 0] 중에서 1, 2, 3이 참이므로 True를 돌려줌.

In [6]:
any([1, 2, 3, 0])

True

리스트 자료형 [0, ""]의 요소 0과 ""은 모두 거짓이므로 False

In [7]:
any([0, ""])

False

## 4. chr
<hr>
chr(i)는 아스키(ASCII) 코드 값을 입력받아 그 코드에 해당하는 문자를 출력

> 아스키 코드란 0에서 127 사이의 숫자를 각각 하나의 문자 또는 기호에 대응시켜 놓은 것

In [8]:
chr(97)

'a'

In [9]:
chr(48)

'0'

## 5. dir
<hr>
dir은 객체가 자체적으로 가지고 있는 변수나 함수를 보여 줌.

In [10]:
dir([1, 2, 3])

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__delitem__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__reversed__',
 '__rmul__',
 '__setattr__',
 '__setitem__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'append',
 'clear',
 'copy',
 'count',
 'extend',
 'index',
 'insert',
 'pop',
 'remove',
 'reverse',
 'sort']

In [11]:
dir({'1': 'a'})

['__class__',
 '__contains__',
 '__delattr__',
 '__delitem__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setitem__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'clear',
 'copy',
 'fromkeys',
 'get',
 'items',
 'keys',
 'pop',
 'popitem',
 'setdefault',
 'update',
 'values']

## 6. divmod
<hr>
divmod(a, b)는 2개의 숫자를 입력 받음. a를 b로 나눈 몫과 나머지를 튜플 형태로 반환

In [12]:
divmod(7, 3)

(2, 1)

In [13]:
7 // 3

2

In [14]:
7 % 3

1

## 7. enumerate
<hr>
enumerate는 "열거하다"라는 뜻. 순서가 있는 자료형(리스트, 튜플, 문자열)을 입력받아 인덱스 값을 포함하는 enumerate 객체를 돌려줌

> 보통 enumerate 함수는 for문과 함께 자주 사용

In [15]:
for i, name in enumerate(['body', 'foo', 'bar']):
    print(i, name)

0 body
1 foo
2 bar


## 8. eval
<hr>
eval(expression)은 실행 가능한 문자열('1+2', "'hi' + 'a'" 같은 것)을 입력으로 받아 문자열을 실행한 결과값을 돌려주는 함수<br>

보통 eval은 입력받은 __문자열__로 파이썬 함수나 클래스를 동적으로 실행하고 싶을 때 사용

In [16]:
eval('1+2')

3

In [20]:
eval("'hi' + 'a'")

'hia'

In [21]:
eval('divmod(4, 3)')

(1, 1)

## 9. filter
<hr>
filter 함수는 첫번째 인수로 함수 이름을, 두번째 인수로 그 함수에 차례로 들어갈 반복 가능한 자료형을 받음. 그리고 두번째 인수인 반복 가능한 자료형 요소가 첫번째 인수인 함수에 입력되었을 때 반환 값이 참인 것만 묶에서 돌려준다.

In [22]:
# positive.py
def positive(l):
    result = []
    for i in l:
        if i > 0:
            result.append(i)
    return result

print(positive([1, -3, 2, 0, -5, 6]))

[1, 2, 6]


위 함수를 filter 함수를 쓰면 간단하게 작성 가능

In [23]:
# filter1.py
def positive(x):
    return x > 0

print(list(filter(positive, [1, -3, 2, 0, -5, 6])))

[1, 2, 6]


## 10. hex
<hr>

hex(x)는 정수 값을 입력받아 16진수(hexadecimal)로 변환하여 돌려주는 함수

In [24]:
hex(234)

'0xea'

In [25]:
hex(3)

'0x3'

## 11. id
<hr>
id(object)는 객체를 입력받아 객체의 고유 주소값(레퍼런스)을 돌려주는 함수

In [26]:
a = 3
id(3)

140708821508560

In [27]:
id(a)

140708821508560

In [28]:
b = a
id(b)

140708821508560

위 예의 3, a, b는 고유 주소값이 모두 140708821508560이다. 즉 3, a, b가 모두 같은 객체를 가리키고 있음. <br><br>
만약 id(4)라고 입력하면 4는 3, a, b와 다른 객체이므로 당연히 다른 고유 주소 값이 출력됨.

In [29]:
id(4)

140708821508592

## 12. input
<hr>
input([prompt])은 사용자 입력을 받는 함수. 매개변수로 문자열을 주면 그 문자열은 프롬프트가 된다

> [] 기호는 괄호 안의 내용을 생략할 수 있다는 관례 표기법

In [30]:
a = input()

hi


In [31]:
a

'hi'

In [32]:
b = input("문자열을 입력하세요: ")

문자열을 입력하세요: 안녕하세요.


In [33]:
b

'안녕하세요.'

## 13. int
<hr>
int(x)는 문자열 형태의 숫자나 소수점이 있는 숫자 등을 정수 형태로 돌려줌.

In [34]:
int('3')

3

In [35]:
int(3.4)

3

int(x, radix)는 radix 진수로 표현된 문자열 x를 10진수로 변환하여 돌려줌.<br><br>
2진수로 표현된 11의 10진수 값은 다음과 같이 구함.

In [36]:
int('11', 2)

3

16진수로 표현된 1A의 10진수 값은 다음과 같이 구함

In [37]:
int('1A', 16)

26

## 14. isinstance
<hr>
isinstance(object, class)는 첫번째 인수로 인스턴스, 두번째 인수로 클래스 이름을 받음. 입력된 인스턴스가 그 클래스의 인스턴스인지 판단

In [38]:
class Person: pass

a = Person()
isinstance(a, Person)

True

In [39]:
b = 3
isinstance(b, Person)

False

## 15. len

In [40]:
len("Python")

6

In [41]:
len([1,2,3])

3

In [42]:
len((1, 'a'))

2

## 16. list
<hr>
list(s)는 반복 가능한 자료형 s를 입력받아 리스트로 만들어 돌려주는 함수.

In [43]:
list("python")

['p', 'y', 't', 'h', 'o', 'n']

In [44]:
list((1,2,3))

[1, 2, 3]

list 함수에 리스트를 입력으로 주면 똑같은 리스트를 복사하여 돌려준다.

In [45]:
a = [1,2,3]
b = list(a)
b

[1, 2, 3]

## 17. map
<hr>
map(f, iterable)은 함수(f)와 반복 가능한(iterable) 자료형을 입력 받음. map은 입력받은 자료형의 각 요소를 함수 f가 수행한 결과를 묶어서 돌려주는 함수.

In [46]:
def two_times(numberList):
    result = []
    for number in numberList:
        result.append(number*2)
    return result

result = two_times([1,2,3,4])
print(result)

[2, 4, 6, 8]


위 예제는 map 함수를 사용하면 다음처럼 바꿀 수 있음.

In [47]:
def two_times(x):
    return x*2

list(map(two_times, [1, 2, 3, 4]))

[2, 4, 6, 8]

lambda 함수를 사용하면 더 간략하게 만들 수 있음.

In [48]:
list(map(lambda a: a*2, [1,2,3,4]))

[2, 4, 6, 8]

## 18. max
<hr>
max(iterable)은 인수로 반복 가능한 자료형을 입력받아 그 최대값을 돌려주는 함수

In [49]:
max([1,2,3])

3

In [50]:
max("python")

'y'

## 19. min
<hr>
min(iterable)은 max 함수와 반대로, 반복 가능한 자료형을 입력받아 그 최소값을 돌려주는 함수

In [51]:
min([1,2,3])

1

In [52]:
min("python")

'h'

## 20. oct
<hr>
oct(x)는 정수 형태의 숫자를 8진수 문자열로 바꾸어 돌려주는 함수.

In [53]:
oct(34)

'0o42'

In [54]:
oct(12345)

'0o30071'

## 21. open
<hr>
open(filename, [mode])는 "파일이름"과 "읽기 방법"을 입력받아 파일 객체를 돌려주는 함수.<br>
읽기 방법(mode)을 생략하면 기본값인 읽기 전용 모드(r)로 파일 객체를 만들어 돌려준다.
<br>

|mode|설명|
|:---|:---|
|w|쓰기 모드로 파일 열기|
|r|읽기 모드로 파일 열기|
|a|추가 모드로 파일 열기|
|b|바이너리 모드로 파일 열기|

b는 w,r,a와 함께 사용한다.

```
f = open("binary_file", "rb")
```

위 예의 rb는 "바이너리 읽기 모드"를 의미 <br>
다음 예의 fread와 fread2는 동일한 방법

```
fread = open("read_mode.txt", 'r')
fread2 = open("read_mode.txt")
```

다음은 추가 모드(a)로 파일을 여는 예

```
fappend = open("append_mode.txt", 'a')
```

## 22. ord
<hr>
ord(c)는 문자의 아스키 코드 값을 돌려주는 함수.

> ord함수는 chr 함수와 반대이다.

In [56]:
ord('a')

97

In [57]:
ord('0')

48

## 23. pow
<hr>
pow(x,y)는 x의 y 제곱한 결과값을 돌려주는 함수.

In [58]:
pow(2,4)

16

In [59]:
pow(3,3)

27

## 24. range
<hr>
range([start,] stop [,step])는 for문과 함께 자주 사용하는 함수. 이 함수는 입력받은 숫자에 해당하는 범위 값을 반복 가능한 객체로 만들어 돌려준다. <br><br>

### 인수가 하나일 경우
<br>
시작 숫자를 지정해 주지 않으면 range 함수는 0부터 시작.

In [60]:
list(range(5))

[0, 1, 2, 3, 4]

### 인수가 2개일 경우
<br>
인수가 2개면 시작 숫자와 끝 숫자를 나타냄. 다만, 끝 숫자는 해당 범위에 포함되지 않는다는 것에 주의

In [61]:
list(range(5,10))

[5, 6, 7, 8, 9]

### 인수가 3개일 경우
<br>
세번째 인수는 숫자 사이의 거리를 뜻함.

In [62]:
list(range(1, 10, 2))

[1, 3, 5, 7, 9]

In [63]:
list(range(0, -10, -2))

[0, -2, -4, -6, -8]

In [64]:
list(range(0, -10, 2))

[]

## 25. sorted
<hr>
sorted(iterable) 함수는 입력값을 정렬할 후 그 결과를 리스트로 돌려줌.

In [65]:
sorted([3,1,2])

[1, 2, 3]

In [66]:
sorted(['a', 'c', 'b'])

['a', 'b', 'c']

In [67]:
sorted("zero")

['e', 'o', 'r', 'z']

In [68]:
sorted((3,2,1))

[1, 2, 3]

리스트 자료형에도 sort 함수가 있음. 하지만 리스트 자료형의 sort함수는 리스트 객체 그 자체를 정렬만 할 뿐 정렬된 결과를 돌려주진 않음.

## 26. str
<hr>
str(object)은 문자열 형태로 객체를 변환하여 돌려주는 함수

In [69]:
str(3)

'3'

In [70]:
str('hi')

'hi'

In [71]:
str('hi'.upper())

'HI'

## 27. sum
<hr>
sum(iterable)은 입력받은 리스트나 튜플의 모든 요소의 합을 돌려주는 함수.

In [72]:
sum([1,2,3])

6

In [73]:
sum((4,5,6))

15

## 28. tuple
<hr>
tuple(iterable)은 반복 가능한 자료형을 입력받아 튜플 형태로 바꿔 돌려줌.

In [74]:
tuple("abc")

('a', 'b', 'c')

In [75]:
tuple([1,2,3])

(1, 2, 3)

In [76]:
tuple((1,2,3))

(1, 2, 3)

## 29. type
<hr>
type(object)은 입력값의 자료형이 무엇인지 알려 주는 함수.

In [77]:
type("abc")

str

In [78]:
type([])

list

In [79]:
type(open("test", 'w'))

_io.TextIOWrapper

## 30. zip
<hr>
zip(*iterable)은 동일한 개수로 이루어진 자료형을 묶어 주는 역할을 하는 함수

> 여기서 사용한 *iterable은 반복 가능(iterable)한 자료형 여러 개를 입력할 수 있다는 의미.

In [80]:
list(zip([1,2,3], [4,5,6]))

[(1, 4), (2, 5), (3, 6)]

In [81]:
list(zip([1,2,3], [4,5,6], [7,8,9]))

[(1, 4, 7), (2, 5, 8), (3, 6, 9)]

In [82]:
list(zip("abc", "defg"))

[('a', 'd'), ('b', 'e'), ('c', 'f')]